runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9270,6 +9270,427 @@ static TR::Register* inlineStringHashCode(TR::Node* node, bool isCompressed, TR:
+   TR::Register *index = cg->allocateRegister();
+   TR::Register *result = cg->allocateRegister();
+   TR::Register *tmp = cg->allocateRegister();
+
+   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions(0, 6, cg);
+
+   deps->addPostCondition(result, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(address, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(index, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(tmp, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(initHash, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(length, TR::RealRegister::NoReg, cg);
+
+   if (!nodeHash)
+      {
+      // If nodeHash is not provided, assume initial hash value of 0.
+      generateRegRegInstruction(TR::InstOpCode::XOR4RegReg, node, initHash, initHash, cg);
+      }
+
+   // Set index ptr to 0
+   generateRegRegInstruction(TR::InstOpCode::XOR4RegReg, node, index, index, cg);
+
+   // Generate Main Loop; 4x Unrolled seems to yield the best performance for large arrays
+   static char *unrollVar = feGetEnv("TR_setInlineVectorHashCodeUnrollCount");
+   int32_t unrollCount = unrollVar ? atoi(unrollVar) : 4;
+
+   vectorizedHashCodeLoopHelper(node, dt, vl, isSigned, result, initHash, index, length, address, unrollCount, cg);
+
+   static bool disableSecondLoop = feGetEnv("TR_disableVectorHashCodeSecondLoop") != NULL;
+
+   // Generate a second vectorized loop;
+   if (!disableSecondLoop)
+      {
+      generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, initHash, result, cg);
+      vectorizedHashCodeLoopHelper(node, dt, TR::VectorLength128, isSigned, result, initHash, index, length, address, 1, cg);
+      }
+
+   // handle residual elements sequentially
+   // for (; index < length; index++) { hash = 31 * hash + arr[index]; }
+   {
+   TR::LabelSymbol *residueBeginLoopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residueEndLoopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residueLoopLabel = generateLabelSymbol(cg);
+
+   residueBeginLoopLabel->setStartInternalControlFlow();
+   residueBeginLoopLabel->setEndInternalControlFlow();
+
+   generateLabelInstruction(TR::InstOpCode::label, node, residueBeginLoopLabel, cg);
+   generateRegRegInstruction(TR::InstOpCode::CMP4RegReg, node, index, length, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE4, node, residueEndLoopLabel, cg);
+   generateLabelInstruction(TR::InstOpCode::label, node, residueLoopLabel, cg);
+
+   // hash = 31 * hash + arr[index] = tmp * hash + arr[index]
+   generateRegRegImmInstruction(TR::InstOpCode::IMULRegRegImm4(), node, result, result, 31, cg);
+
+   static TR::InstOpCode::Mnemonic signedLoadOpcode[3]   = { TR::InstOpCode::MOVSXReg4Mem1, TR::InstOpCode::MOVSXReg4Mem2, TR::InstOpCode::L4RegMem };
+   static TR::InstOpCode::Mnemonic unsignedLoadOpcode[3] = { TR::InstOpCode::L1RegMem, TR::InstOpCode::L2RegMem, TR::InstOpCode::L4RegMem };
```

Comment:
Unsigned load opcodes should use MOVZXReg4Mem1/MOVZXReg4Mem2/L4RegMem, not 1/2 byte loads because it does not zero out the high bits which could mess up the ADD.